`hvg`  selects more genes than asked for

### Please make sure these conditions are met

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the latest version of scanpy.
- [X] (optional) I have confirmed this bug exists on the main branch of scanpy.

### What happened?

HVG can produce more than the number of genes asked for as highly variable.

This occurs on these two datasets:

```
wget https://datasets.cellxgene.cziscience.com/e00ab1f4-28cd-497d-b889-94d45840f423.h5ad
```

### Minimal code sample

```python
import scanpy as sc

adata1 = sc.read('e00ab1f4-28cd-497d-b889-94d45840f423.h5ad')

sc.pp.normalize_total(adata1, target_sum=1e4)

sc.pp.log1p(adata1)

n_top_gene = 10000
sc.pp.highly_variable_genes(adata1, n_top_genes = n_top_gene)

hvg_system1 = set(adata1.var[adata1.var['highly_variable']].index)
assert len(hvg_system1) == n_top_gene, f"found {len(hvg_system1)} instead of {n_top_gene}"

```


### Error output

```pytb
AssertionError                            Traceback (most recent call last)
Cell In[12], line 1
----> 1 assert len(hvg_system1) == n_top_gene, f"found {len(hvg_system1)} instead of {n_top_gene}"

AssertionError: found 13355 instead of 10000
```


### Versions

<details>

```
import scanpy; scanpy.logging.print_versions()
-----
anndata     0.10.8
scanpy      1.10.0rc2.dev85+gb918a23e
-----
```

</details>

--- a/docs/release-notes/1.10.3.md
+++ b/docs/release-notes/1.10.3.md
@@ -11,5 +11,6 @@
 * Fix `subset=True` of {func}`~scanpy.pp.highly_variable_genes` when `flavor` is `seurat` or `cell_ranger`, and `batch_key!=None` {pr}`3042` {smaller}`E Roellin`
 * Add compatibility with {mod}`numpy` 2.0 {pr}`3065` and {pr}`3115` {smaller}`P Angerer`
 * Fix `legend_loc` argument in {func}`scanpy.pl.embedding` not accepting matplotlib parameters {pr}`3163` {smaller}`P Angerer`
+* Fix dispersion cutoff in {func}`~scanpy.pp.highly_variable_genes` in presence of `NaN`s {pr}`3176` {smaller}`P Angerer`
 
 #### Performance

--- a/src/scanpy/preprocessing/_highly_variable_genes.py
+++ b/src/scanpy/preprocessing/_highly_variable_genes.py
@@ -402,7 +402,7 @@ def _subset_genes(
         f"the {n_top_genes} top genes correspond to a "
         f"normalized dispersion cutoff of {disp_cut_off}"
     )
-    return np.nan_to_num(dispersion_norm) >= disp_cut_off
+    return np.nan_to_num(dispersion_norm, nan=-np.inf) >= disp_cut_off
 
 
 def _nth_highest(x: NDArray[np.float64] | DaskArray, n: int) -> float | DaskArray:

--- a/tests/test_highly_variable_genes.py
+++ b/tests/test_highly_variable_genes.py
@@ -13,6 +13,7 @@ from pandas.testing import assert_frame_equal, assert_index_equal
 from scipy import sparse
 
 import scanpy as sc
+from scanpy.preprocessing._utils import _get_mean_var
 from testing.scanpy._helpers import _check_check_values_warnings
 from testing.scanpy._helpers.data import pbmc3k, pbmc68k_reduced
 from testing.scanpy._pytest.marks import needs
@@ -129,6 +130,27 @@ def test_keep_layer(base, flavor):
         assert False
 
     assert np.allclose(X_orig.toarray(), adata.X.toarray())
+
+
+@pytest.mark.parametrize(
+    "flavor",
+    [
+        "seurat",
+        pytest.param(
+            "cell_ranger",
+            marks=pytest.mark.xfail(reason="can’t deal with duplicate bin edges"),
+        ),
+    ],
+)
+def test_no_filter_genes(flavor):
+    """Test that even with columns containing all-zeros in the data, n_top_genes is respected."""
+    adata = sc.datasets.pbmc3k()
+    means, _ = _get_mean_var(adata.X)
+    assert (means == 0).any()
+    sc.pp.normalize_total(adata, target_sum=10000)
+    sc.pp.log1p(adata)
+    sc.pp.highly_variable_genes(adata, flavor=flavor, n_top_genes=10000)
+    assert adata.var["highly_variable"].sum() == 10000
 
 
 def _check_pearson_hvg_columns(output_df: pd.DataFrame, n_top_genes: int):

